google expects plain schema-org url

pvgenuchten opened the pull request:
This PR started out as a 'simple' namespace redefine. However it turns out to be related to a wider issue on the schema-org-as-json-ld domain. https://schema.org itself does not advertise a ld-context, but crawlers still require this url to be used as namespace. pyld, which is used to validate the json-ld, however requires a valid ld-context.

That's why the initial commit was reverted and a hacky-workaround is suggested, within the ajax-response the schema-org url is rewritten to fit the needs of search engine crawlers.

Issue does resolve #574

tomkralidis commented:
+1 from me, thanks @pvgenuchten

alpha-beta-soup commented:
This implies reverting to the original, that had to be changed for _reasons_ https://github.com/geopython/pygeoapi/issues/444 and these reasons still apply (that's why the tests failed). It's still an upstream problem with pyld.

alpha-beta-soup reviewed (changes requested):
This change causes tests to fail, since they depend on pyld to parse the JSON-LD metadata. Either we fix pyld upstream, or use a different Python JSON-LD parsing library (or perhaps don't parse it at all).

pvgenuchten commented:
i was afraid of this... 
Or fix the schema-org validator...
as a workaround, maybe we can shorten the context url as part of the ajax request that introduces the snippet?

pvgenuchten commented:
just tested with google webmaster console
that tool seems to trigger only on the microdata and not on the embedded json-ld (in contradiction to structured data testing tool)
it could be related to the schem-org context url, but a bit puzzled/worried here about the impact of this

dblodgett-usgs commented:
@ksonda -- do you have any recollection of the nuances here?

ksonda commented:
I'm not familiar with pyld or why that would be affected, but Google will be retiring the structured data test tool, which didn't seem to pull in remote contexts other than schema.org. You could specify predicates manually though. The new tool, Google Rich Results, [will only parse microdata for resources of the following schema.org types](https://support.google.com/webmasters/answer/7445569):

![image](https://user-images.githubusercontent.com/44071350/99681372-100b7700-2a4c-11eb-948e-78bbaa11c84d.png)

pvgenuchten commented:
hi @dblodgett-usgs, 

nice to see the demo based on master is redeployed with the microdata being removed from last merged PR... helps testing

The json-ld is now only [injected using ajax](https://github.com/geopython/pygeoapi/blob/4e5cfd8e50ad760d8f2a958645ac0cb256d88236/pygeoapi/templates/base.html#L62-L73). 

I noticed structured-data-testing tool picks up the injected script code. 

![image](https://user-images.githubusercontent.com/299829/99681247-36e2a100-2a7e-11eb-8666-1419341b49c1.png)

However [rich results test](https://search.google.com/test/rich-results) does not find any embedded structured data.

![image](https://user-images.githubusercontent.com/299829/99681097-10246a80-2a7e-11eb-8e56-c5e313c8ea5d.png)

Also yandex is not able to retrieve any 

![image](https://user-images.githubusercontent.com/299829/99681707-bbcdba80-2a7e-11eb-8a5d-92dc58cad470.png)

I guess the ajax approach is not optimal, at least for some clients. Do we have an option to embed the json-ld using python?

some reading:
https://webmasters.stackexchange.com/questions/100935/why-wouldnt-google-be-able-to-read-ajax-generated-json-ld-schema-org-markup-par

pvgenuchten commented:
Note that the some of the problems may also be caused by https://github.com/geopython/pygeoapi/issues/577,

indeed:

![image](https://user-images.githubusercontent.com/299829/99693660-ea05c700-2a8b-11eb-9da1-2f09f87c45bc.png)

but still curious why it finds a button and not a dataset

pvgenuchten commented:
I implemented the following test scenario:
- i made a basic html page available at http://pygeoapi.genuchten.net
- the page ingests schema org using ajax similar as pygeoapi
- i tested the page in google search console
- i updated context url to be plain schema.org
result: schema.org is parsed

![image](https://user-images.githubusercontent.com/299829/100008876-4dfef700-2dce-11eb-8e5e-57beb2c21309.png)

- i updated the context to the [full context path](http://pygeoapi.genuchten.net/full.html)

then google complains:

![image](https://user-images.githubusercontent.com/299829/100009197-d4b3d400-2dce-11eb-92fa-5efecd050f3d.png)

alpha-beta-soup commented:
I still think the correct option is to attempt a PR to pyld so that it is able to use the appropriate schema.org URL; this would benefit other projects, and avoids the inconsistency between the direct JSON-LD response and the embedded data.

alpha-beta-soup commented:
https://github.com/digitalbazaar/pyld/pull/129

pvgenuchten commented:
@alpha-beta-soup i agree with the direction, however currently schema.org is broken since the removal of schema-org-microdata, i see this as a workaround until pyld would be optimised, are you ok to adopt it as a temporary workaround? Alternative option would be to revert the microdata removal.

alpha-beta-soup commented:
Not sure it's my call. My preference would be to keep the microdata, but I'm OK with the "temporary workaround" too as long as there's a corresponding issue/s that so it doesn't drop off the radar.

pvgenuchten commented:
@tomkralidis this is ready to be merged (fixing schema-org in crawlers), while keeping open the discussion with pyld

pvgenuchten commented:
thanx Tom
